File: backend/business_core/strategy/twelve_week_year.py
from datetime import datetime
from typing import Optional, List

from sqlalchemy import String, Boolean, DateTime, ForeignKey, BigInteger, func, UniqueConstraint, Text, Integer, Numeric, Float, Index, text
from sqlalchemy.dialects.postgresql import JSONB, TSVECTOR
from sqlalchemy.orm import Mapped, mapped_column, relationship, object_session
from pgvector.sqlalchemy import Vector

from db.base_class import Base
from core.snowflake import generate_snowflake_id

class TwelveWeekCycle(Base):
    __tablename__ = "twelve_week_cycles"
    __table_args__ = (
        # §6.2: "twelve_week_cycles: unique (brain_id, start_date)".
        UniqueConstraint('brain_id', 'start_date', name='uix_twelve_week_cycle_brain_start'),
        Index("ix_twelve_week_cycle_mvp_stage_id", "mvp_stage_id"),
        Index("ix_twelve_week_cycle_project_id", "project_id"),
        {"schema": "operating"},
    )
    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    brain_id: Mapped[int] = mapped_column(ForeignKey("knowledge.brains.id"), index=True)
    project_id: Mapped[Optional[int]] = mapped_column(ForeignKey("strategy.projects.id"), nullable=True, index=True)
    # References MvpStage, not the unrelated CycleStage - see WeeklyPlan.stage_id.
    mvp_stage_id: Mapped[Optional[int]] = mapped_column(ForeignKey("strategy.mvp_stages.id"), nullable=True)
    okr_cycle_id: Mapped[Optional[int]] = mapped_column(ForeignKey("strategy.okr_cycles.id"), nullable=True)
    cycle_contract_id: Mapped[Optional[int]] = mapped_column(ForeignKey("operating.cycle_contracts.id", use_alter=True), nullable=True, index=True)
    theme: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    vision_statement: Mapped[str] = mapped_column(Text, default="")
    stage_at_start: Mapped[str] = mapped_column(String(50), default="S1_PROBLEM_VALIDATION")
    current_week: Mapped[int] = mapped_column(Integer, default=1)
    duration_weeks: Mapped[int] = mapped_column(Integer, default=12)
    overall_execution_score: Mapped[float] = mapped_column(Float, default=0.0)
    start_date: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    end_date: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    commitment_level: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="ACTIVE")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class WeeklyPlan(Base):
    __tablename__ = "weekly_plans"
    __table_args__ = (
        # §6.2: "weekly_plans: unique (cycle_id, week_no) với week_no từ 1 đến 12" -
        # thiếu ràng buộc này thì 2 tuần trùng số thứ tự trong cùng 1 cycle vẫn được
        # tạo bình thường, phá vỡ đúng bất biến 12-Week Year.
        UniqueConstraint('cycle_id', 'week_no', name='uix_weekly_plan_cycle_week'),
        {"schema": "operating"},
    )
    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id"), index=True)
    stage_id: Mapped[Optional[int]] = mapped_column(ForeignKey("operating.cycle_stages.id"), nullable=True, index=True)
    week_no: Mapped[int] = mapped_column()
    start_date: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    end_date: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    focus: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    mission: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    success_criteria: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    execution_score: Mapped[Optional[float]] = mapped_column(nullable=True)
    outcome_score: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    blockers_jsonb: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    reflection: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class WeeklyCommitment(Base):
    __tablename__ = "weekly_commitments"
    __table_args__ = {"schema": "operating"}
    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    weekly_plan_id: Mapped[int] = mapped_column(ForeignKey("operating.weekly_plans.id"), index=True)
    initiative_id: Mapped[Optional[int]] = mapped_column(ForeignKey("strategy.initiatives.id"), nullable=True)
    title: Mapped[str] = mapped_column(String(255))
    status: Mapped[str] = mapped_column(String(50), default="todo")
    planned_effort: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    commitment_owner_type: Mapped[Optional[str]] = mapped_column(String(50), default="FOUNDER")  # FOUNDER, AI_AGENT, CORE_TEAM
    execution_mode: Mapped[Optional[str]] = mapped_column(String(50), default="MANUAL")  # MANUAL, AI_ASSISTED, AUTONOMOUS
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


# ==========================================
# mCOSA V12 — Project Intelligence & 12WY Additions
# ==========================================

class CycleContract(Base):
    __tablename__ = "cycle_contracts"
    __table_args__ = {"schema": "operating"}
    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id", ondelete="CASCADE"), unique=True, index=True)
    success_definition: Mapped[str] = mapped_column(Text)
    goal_ids: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    kr_ids: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    founder_capacity_per_week: Mapped[Optional[float]] = mapped_column(Float, nullable=True)  # hours/week
    reserved_buffer_percent: Mapped[Optional[float]] = mapped_column(Float, nullable=True)  # e.g. 20.0
    ai_budget: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    operating_budget: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    risk_constraints: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="draft")  # draft, approved, active
    approved_by: Mapped[Optional[int]] = mapped_column(ForeignKey("core.users.id"), nullable=True)
    approved_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class CycleStage(Base):
    __tablename__ = "cycle_stages"
    __table_args__ = (
        UniqueConstraint('cycle_id', 'order_no', name='uix_cycle_stage_cycle_order'),
        {"schema": "operating"},
    )
    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(100))  # Discovery, Validation, MVP, Beta, Acquisition, Closing, Week 13 Review
    purpose: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    start_week: Mapped[int] = mapped_column(Integer)
    end_week: Mapped[int] = mapped_column(Integer)
    order_no: Mapped[int] = mapped_column(Integer, default=1)
    expected_outcomes: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="pending")  # pending, in_progress, completed, skipped
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class WeeklyReview(Base):
    """Bản đánh giá tuần (Weekly Review - Spec §17).

    Ghi nhận bằng chứng đã học, giả định được xác nhận/bác bỏ, và khuyến nghị hành động.
    """
    __tablename__ = "weekly_reviews"
    __table_args__ = (
        UniqueConstraint("weekly_plan_id", name="uix_weekly_review_plan"),
        {"schema": "operating"},
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id"), index=True)
    weekly_plan_id: Mapped[int] = mapped_column(ForeignKey("operating.weekly_plans.id"), index=True)
    week_no: Mapped[int] = mapped_column()
    execution_score: Mapped[float] = mapped_column(Float, default=0.0)
    outcome_score: Mapped[float] = mapped_column(Float, default=0.0)
    evidence_learned: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    assumptions_confirmed: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    assumptions_invalidated: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    recommendation: Mapped[str] = mapped_column(String(50), default="CONTINUE")  # CONTINUE, PIVOT_NEXT_WEEK, DOUBLE_DOWN, RECALIBRATE_CAPACITY
    narrative_summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    reviewed_by: Mapped[int] = mapped_column(ForeignKey("core.users.id"), index=True)
    reviewed_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class CycleReview(Base):
    """Bản tổng kết chu kỳ 12 tuần (Cycle Retrospective & Review - Spec §18)."""
    __tablename__ = "cycle_reviews"
    __table_args__ = (
        UniqueConstraint("cycle_id", name="uix_cycle_review_cycle"),
        {"schema": "operating"},
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id"), index=True)
    overall_execution_score: Mapped[float] = mapped_column(Float, default=0.0)
    overall_outcome_score: Mapped[float] = mapped_column(Float, default=0.0)
    okr_achievement_rate: Mapped[float] = mapped_column(Float, default=0.0)
    strategic_learnings: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    systemic_blockers: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    portfolio_adjustments: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    next_cycle_recommendations: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="finalized")
    reviewed_by: Mapped[int] = mapped_column(ForeignKey("core.users.id"), index=True)
    reviewed_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class CelebrationRecord(Base):
    """Ghi nhận lễ kỷ niệm & vinh danh thành tựu chu kỳ (Week 13 Celebration - Spec §19)."""
    __tablename__ = "celebration_records"
    __table_args__ = {"schema": "operating"}

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id"), index=True)
    title: Mapped[str] = mapped_column(String(255))
    milestones_achieved: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    top_performers_recognized: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)
    rewards_or_rituals: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    reflection_notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_by: Mapped[int] = mapped_column(ForeignKey("core.users.id"), index=True)
    celebrated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


# ==========================================
# mCOSA V12 — Portfolio Intelligence (Sprint 6)
# ==========================================

class TacticalExecutionItem(Base):
    """Hành động chiến thuật tuần & Chỉ số dẫn dắt (12WY Tactics & Lead Indicators)."""
    __tablename__ = "tactical_execution_items"
    __table_args__ = {"schema": "operating"}
    
    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("strategy.projects.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id"), index=True)
    
    week_number: Mapped[int] = mapped_column(Integer, index=True) # 1 -> 12
    title: Mapped[str] = mapped_column(String(255))
    description: Mapped[str] = mapped_column(Text, default="")
    
    tows_option_id: Mapped[Optional[int]] = mapped_column(ForeignKey("strategy.tows_options.id"), nullable=True)
    hypothesis_id: Mapped[Optional[int]] = mapped_column(ForeignKey("strategy.hypotheses.id"), nullable=True)
    
    lead_indicator_name: Mapped[str] = mapped_column(String(255))
    target_count: Mapped[int] = mapped_column(Integer, default=1)
    actual_count: Mapped[int] = mapped_column(Integer, default=0)
    
    status: Mapped[str] = mapped_column(String(50), default="PLANNED") # PLANNED | IN_PROGRESS | DONE | BLOCKED
    owner_role: Mapped[str] = mapped_column(String(100), default="Founder")
    
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

class WeeklyAccountabilityReview(Base):
    """Nhật ký phiên kiểm điểm tiến độ tuần (Weekly Accountability Meeting - WAM)."""
    __tablename__ = "weekly_accountability_reviews"
    __table_args__ = {"schema": "operating"}
    
    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=False, default=generate_snowflake_id)
    workspace_id: Mapped[int] = mapped_column(ForeignKey("core.workspaces.id"), index=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("strategy.projects.id"), index=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("operating.twelve_week_cycles.id"), index=True)
    
    week_number: Mapped[int] = mapped_column(Integer)
    execution_score: Mapped[float] = mapped_column(Float, default=0.0)
    
    total_planned: Mapped[int] = mapped_column(Integer, default=0)
    total_completed: Mapped[int] = mapped_column(Integer, default=0)
    
    key_breakthroughs: Mapped[dict] = mapped_column(JSONB, default=list)
    root_cause_blocks: Mapped[dict] = mapped_column(JSONB, default=list)
    ai_recommendations: Mapped[dict] = mapped_column(JSONB, default=list)
    
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

File: services/operations/twelve-week-year.ts
import { api, APIError } from "encore.dev/api";
import { SQLDatabase } from "encore.dev/storage/sqldb";

const db = SQLDatabase.named("operations");

export interface TwelveWeekCycle {
  id: number;
  workspaceId: number;
  brainId?: number | null;
  projectId?: number | null;
  theme?: string | null;
  visionStatement: string;
  stageAtStart: string;
  currentWeek: number;
  durationWeeks: number;
  overallExecutionScore: number;
  startDate?: string | null;
  endDate?: string | null;
  commitmentLevel?: string | null;
  status: string;
  createdAt: string;
}

export interface CreateTwelveWeekCycleRequest {
  workspaceId: number;
  brainId?: number | null;
  projectId?: number | null;
  theme?: string | null;
  visionStatement?: string;
  stageAtStart?: string;
  durationWeeks?: number;
  startDate?: string | null;
  endDate?: string | null;
  commitmentLevel?: string | null;
}

export interface WeeklyPlan {
  id: number;
  workspaceId: number;
  cycleId: number;
  weekNo: number;
  startDate?: string | null;
  endDate?: string | null;
  focus?: string | null;
  mission?: string | null;
  executionScore?: number | null;
  outcomeScore?: number | null;
  reflection?: string | null;
  createdAt: string;
}

export interface CreateWeeklyPlanRequest {
  workspaceId: number;
  cycleId: number;
  weekNo: number;
  startDate?: string | null;
  endDate?: string | null;
  focus?: string | null;
  mission?: string | null;
}

export interface WeeklyCommitment {
  id: number;
  workspaceId: number;
  weeklyPlanId: number;
  initiativeId?: number | null;
  title: string;
  status: string;
  plannedEffort?: string | null;
  commitmentOwnerType?: string | null;
  executionMode?: string | null;
  createdAt: string;
}

export interface CreateWeeklyCommitmentRequest {
  workspaceId: number;
  weeklyPlanId: number;
  initiativeId?: number | null;
  title: string;
  plannedEffort?: string | null;
  commitmentOwnerType?: string | null;
  executionMode?: string | null;
}

// ─── 12-Week Cycles Endpoints ───

export const createCycle = api(
  { expose: true, method: "POST", path: "/operations/cycles" },
  async (req: CreateTwelveWeekCycleRequest): Promise<TwelveWeekCycle> => {
    if (!req.workspaceId) throw APIError.invalidArgument("workspaceId is required");

    const row = await db.queryRow<TwelveWeekCycle>`
      INSERT INTO operating.twelve_week_cycles (
        workspace_id, brain_id, project_id, theme, vision_statement,
        stage_at_start, duration_weeks, start_date, end_date, commitment_level
      ) VALUES (
        ${req.workspaceId}, ${req.brainId ?? null}, ${req.projectId ?? null},
        ${req.theme ?? null}, ${req.visionStatement ?? ""},
        ${req.stageAtStart ?? "S1_PROBLEM_VALIDATION"}, ${req.durationWeeks ?? 12},
        ${req.startDate ?? null}, ${req.endDate ?? null}, ${req.commitmentLevel ?? null}
      )
      RETURNING
        id, workspace_id as "workspaceId", brain_id as "brainId", project_id as "projectId",
        theme, vision_statement as "visionStatement", stage_at_start as "stageAtStart",
        current_week as "currentWeek", duration_weeks as "durationWeeks",
        overall_execution_score as "overallExecutionScore",
        start_date as "startDate", end_date as "endDate",
        commitment_level as "commitmentLevel", status, created_at as "createdAt"
    `;
    if (!row) throw APIError.internal("Failed to create twelve week cycle");
    return row;
  }
);

export const listCycles = api(
  { expose: true, method: "GET", path: "/operations/workspaces/:workspaceId/cycles" },
  async (params: { workspaceId: number }): Promise<{ cycles: TwelveWeekCycle[] }> => {
    const rows = db.query<TwelveWeekCycle>`
      SELECT
        id, workspace_id as "workspaceId", brain_id as "brainId", project_id as "projectId",
        theme, vision_statement as "visionStatement", stage_at_start as "stageAtStart",
        current_week as "currentWeek", duration_weeks as "durationWeeks",
        overall_execution_score as "overallExecutionScore",
        start_date as "startDate", end_date as "endDate",
        commitment_level as "commitmentLevel", status, created_at as "createdAt"
      FROM operating.twelve_week_cycles
      WHERE workspace_id = ${params.workspaceId}
      ORDER BY id DESC
    `;
    const cycles: TwelveWeekCycle[] = [];
    for await (const row of rows) cycles.push(row);
    return { cycles };
  }
);

// ─── Weekly Plans Endpoints ───

export const createWeeklyPlan = api(
  { expose: true, method: "POST", path: "/operations/weekly-plans" },
  async (req: CreateWeeklyPlanRequest): Promise<WeeklyPlan> => {
    if (!req.workspaceId || !req.cycleId || !req.weekNo) {
      throw APIError.invalidArgument("workspaceId, cycleId, and weekNo are required");
    }

    const row = await db.queryRow<WeeklyPlan>`
      INSERT INTO operating.weekly_plans (
        workspace_id, cycle_id, week_no, start_date, end_date, focus, mission
      ) VALUES (
        ${req.workspaceId}, ${req.cycleId}, ${req.weekNo},
        ${req.startDate ?? null}, ${req.endDate ?? null},
        ${req.focus ?? null}, ${req.mission ?? null}
      )
      RETURNING
        id, workspace_id as "workspaceId", cycle_id as "cycleId", week_no as "weekNo",
        start_date as "startDate", end_date as "endDate",
        focus, mission, execution_score as "executionScore",
        outcome_score as "outcomeScore", reflection, created_at as "createdAt"
    `;
    if (!row) throw APIError.internal("Failed to create weekly plan");
    return row;
  }
);

// ─── Weekly Commitments Endpoints ───

export const createWeeklyCommitment = api(
  { expose: true, method: "POST", path: "/operations/weekly-commitments" },
  async (req: CreateWeeklyCommitmentRequest): Promise<WeeklyCommitment> => {
    if (!req.workspaceId || !req.weeklyPlanId || !req.title) {
      throw APIError.invalidArgument("workspaceId, weeklyPlanId, and title are required");
    }

    const row = await db.queryRow<WeeklyCommitment>`
      INSERT INTO operating.weekly_commitments (
        workspace_id, weekly_plan_id, initiative_id, title, planned_effort,
        commitment_owner_type, execution_mode
      ) VALUES (
        ${req.workspaceId}, ${req.weeklyPlanId}, ${req.initiativeId ?? null},
        ${req.title}, ${req.plannedEffort ?? null},
        ${req.commitmentOwnerType ?? "FOUNDER"}, ${req.executionMode ?? "MANUAL"}
      )
      RETURNING
        id, workspace_id as "workspaceId", weekly_plan_id as "weeklyPlanId",
        initiative_id as "initiativeId", title, status,
        planned_effort as "plannedEffort",
        commitment_owner_type as "commitmentOwnerType",
        execution_mode as "executionMode", created_at as "createdAt"
    `;
    if (!row) throw APIError.internal("Failed to create weekly commitment");
    return row;
  }
);
